spec/models/queries/work_packages/filter/ancestor_filter_spec.rb

```diff
@@ -0,0 +1,312 @@
+            .and_return([visible_wp.id])
+
+          instance.valid_values!
+
+          expect(instance.values)
+            .to contain_exactly(visible_wp.id.to_s)
+        end
+      end
+    end
+
+    describe "#validate" do
+      let(:visible_wp) { build_stubbed(:work_package) }
+      let(:invisible_wp) { build_stubbed(:work_package) }
+
+      context "with old templated value" do
+        it "is still valid" do
+          instance.values = %w[templated]
+          expect(instance).to be_valid
+        end
+      end
+
+      context "with new templated value" do
+        it "is still valid" do
+          instance.values = %w[{id}]
+          expect(instance).to be_valid
+        end
+      end
+
+      context "within a project" do
+        it "is valid if only visible wps are values" do
+          instance.values = [visible_wp.id.to_s]
+
+          allow(WorkPackage)
+            .to receive_message_chain(:visible, :for_projects, :where, :pluck) # rubocop:disable RSpec/MessageChain
+            .with(no_args)
+            .with(project)
+            .with(id: instance.values)
+            .with(:id)
+            .and_return([visible_wp.id])
+
+          expect(instance).to be_valid
+        end
+
+        it "is invalid if invisible wps are values" do
+          instance.values = [invisible_wp.id.to_s, visible_wp.id.to_s]
+
+          allow(WorkPackage)
+            .to receive_message_chain(:visible, :for_projects, :where, :pluck) # rubocop:disable RSpec/MessageChain
+            .with(no_args)
+            .with(project)
+            .with(id: instance.values)
+            .with(:id)
+            .and_return([visible_wp.id])
+
+          expect(instance).not_to be_valid
+        end
+      end
+
+      context "when outside of a project" do
+        let(:project) { nil }
+
+        it "is valid if only visible wps are values" do
+          instance.values = [visible_wp.id.to_s]
+
+          allow(WorkPackage)
+            .to receive_message_chain(:visible, :where, :pluck) # rubocop:disable RSpec/MessageChain
+            .with(no_args)
+            .with(id: instance.values)
+            .with(:id)
+            .and_return([visible_wp.id])
+
+          expect(instance).to be_valid
+        end
+
+        it "is invalid if invisible wps are values" do
+          instance.values = [invisible_wp.id.to_s, visible_wp.id.to_s]
+
+          allow(WorkPackage)
+            .to receive_message_chain(:visible, :where, :pluck) # rubocop:disable RSpec/MessageChain
+            .with(no_args)
+            .with(id: instance.values)
+            .with(:id)
+            .and_return([visible_wp.id])
+
+          expect(instance).not_to be_valid
+        end
+      end
+    end
+
+    describe "#where and #includes" do
+      let(:parent) { create(:work_package) }
+      let(:child) { create(:work_package, parent:) }
+      let(:grandchild) { create(:work_package, parent: child) }
+      let(:another_wp) { create(:work_package) }
```

Comment:
I think it would be good to also include a grandparent here which then also shouldn't be returned on `=`.

Reply:
Good idea, I added a grandparent to these specs 👍🏻